### ISSUE TITLE
scopes not working in jwt

### ISSUE THREAD

**Issue by skaugvoll** [felixheck/hapi-auth-keycloak#98]
Hi i',m having trouble using scopes with the jwt and this.
The jwt has in the root a field named scope wich is a string of different scopes, separated by a whitespace.

I've played around in the src code of the library, and it seems like token.js line 50 
`const scopes = _.flatten(_.map(auth.permissions, 'scopes')).map(prefix('scope'))` does not return any scopes. I know the scope is present. If i on line 50-51, add the unpacking og scope variable, and console.log it it shows me my scopes.

The function returns the scope  as an empty list. This is a problem for me. 

If I change line 
```const scopes = _.flatten(_.map(auth.permissions, 'scopes')).map(prefix('scope'))``` to ``` const scopes = scope.split(" ").map(s => `scope:${s}`);``` it works

token.js
```
function getRoles (clientId, {
  realm_access: realm = { roles: [] },
  resource_access: resources = {},
  authorization: auth = { permissions: [] },
  scope
}) {
  const prefix = prefixRole.bind(undefined, clientId)
  const realmRoles = realm.roles.map(prefix('realm'))
  // const scopes = _.flatten(_.map(auth.permissions, 'scopes')).map(prefix('scope'))
  const scopes = scope.split(" ").map(s => `scope:${s}`);
  const appRoles = Object.keys(resources).map((key) => resources[key].roles.map(prefix(key)))
  return _.flattenDepth([realmRoles, scopes, appRoles], 2)
}```

**Comment by felixheck** [comment]
Hi!

Thanks for reporting this issue and using this plugin.
Would you mind sharing how you generated this JWT, which keycloak client settings you've used as well as the keycloak version? Maybe even how you defined the scopes.

Cheers

**Comment by skaugvoll** [comment]
Thanks for responding, I would love to, but the specific keycloak settings and how it is set up is out of my controll. but the protocol is OpenID Connect (OIDC)
Cheers

**Comment by felixheck** [comment]
Let me figure out when keycloak sets the scope as `auth.permissions.[x].scopes` and when as `scope` — maybe the plugin needs to support both in some way.

**Comment by skaugvoll** [comment]
Thank you! for covering most use cases, I would think supporting both of them would be a good idea. What if;
```
const scopesPermissions = _.flatten(_.map(auth.permissions, 'scopes')).map(prefix('scope'))
const scopesRoot = scope.split(" ").map(s => `scope:${s}`);

const scopes = scopesPermissions.concat(scopesRoot)?
```

**Comment by felixheck** [comment]
yep something like that 🤔 
wanna understand first if those scopes differ somehow and need to be treated separately.

Let me reach out to my colleagues

**Comment by felixheck** [comment]
https://github.com/keycloak/keycloak-documentation/blob/master/server_admin/topics/clients/client-scopes.adoc

your scopes seem to be client scopes, right?

**Comment by skaugvoll** [comment]
Yes sir, that is correct. Client scopes and Optional Client scopes for OpenID Connect, according to the link you posted above

**Comment by felixheck** [comment]
If I got it right, client scopes are something that the client itself can pass to Keycloak/OIDC providers to get tokens adapted to the client's use cases.

The scopes covered by the plugin so far, however, describe rather the [permissions](https://www.keycloak.org/docs/5.0/authorization_services/#_permission_create_scope) assigned to the user and are therefore important for authorization. `write:document`,  for example could describe that the user has the permissions to edit a certain resource.

Therefore, the approaches differ fundamentally and I'm not sure how the client scopes can help with authorization. What do you intend to do with it? So can you explain why you'd like to check the client scope with hapi's built-in authorization feature?

**Comment by skaugvoll** [comment]
I want to use the client scopes to restrict permission/access to different API endpoints alongside the validation of the accesstoken (which your library does very well and easy to implement). So to use your library in an OAuth 2.0 resource server manner, it has to support client scopes (at least for my use case). And since a client can have access to only a set of the scopes available, I think it makes sense to include them in the library for authorization reasons ?

My resources:
https://oauth.net/2/scope/
https://openid.net/specs/openid-connect-core-1_0.html#ScopeClaims

**Comment by felixheck** [comment]
thanks for clarifying — how about prefixing them with `clientscope:` to differ clearly between both?

**Comment by skaugvoll** [comment]
you are thinking of this line `scope: ['realm:admin', 'editor', 'other-resource:creator', 'scope:foo.READ']` (got it from your examples), so instead of prefix with `scope:foo.READ`, I would access the scopes I need with `clientscope:foo.READ`

I think that makes sense, as long as it states in the documentation the difference between scope and that the clientscope is for Oauth2.0/OIDC, client scopes

**Comment by felixheck** [comment]
Yep, so existing (permission) scopes will still be prefixed with `scope:` as in the official Keycloak express plugin. I'd like to keep some consistency here.

And the client scopes will be prefixed with `clientscope:` — proper docs are matter of course.
Do you have time & interest in submitting an initial PR regarding this feature with tests etc? I'm currently pretty busy :-/

**Comment by skaugvoll** [comment]
I'm not sure I have  the time either, as it would take me some time to figure out how to write and follow your test setup. But if you don't have the time, I might be able to use a few late nights on it. Let me know :)

**Comment by felixheck** [comment]
another proposal:
I'll prepare a release candidate with the implementation but without tests — so you can start using it and I'll add tests gradually within the next weeks 😊

**Comment by felixheck** [comment]
Published as `5.1.0` – thanks again 😊
